### ISSUE TITLE
SignalR Timeout 

### ISSUE THREAD

**Issue by mandalorianbob** [dotnet/aspnetcore#22687]
I'm having an issue with timeouts when using SignalR, connected to an Azure backplane, on a Blazor WASM frontend.

My use case is that I've got a webpage to assist in playing otherwise in-person games over the internet. To this end, I want someone to be able to open the page, sit there for perhaps many minutes or even hours, occasionally receiving messages about dice rolls that are happening, and then click a button and roll dice.

Unfortunately, the SignalR hub drops the connection after a few minutes of inactivity, even though the web browser is open and on the page.

How is this situation supposed to be handled? Is there any way to increase this timeout - I tried increasing the servertimeout that didn't seem to do it - or handle it more gracefully?

I assume there's some kind of reconnect strategy here - but I tried putting "WithAutomaticReconnect" and that didn't seem to have any effect either.

I'm using Blazor.HttpClient 3.2.0-preview3.20168.3, SignalR.Client 5.0.0-preview.4.20257.10 and Blazor 3.2.0

Here's how I'm building the connections:
```
 _hubConnection = new HubConnectionBuilder()
                 .WithUrl(NavigationManager.ToAbsoluteUri($"/{HubMethods.IfritRollResultsHub.Name}"))
                 .WithAutomaticReconnect()
                 .Build();
            _hubConnection.ServerTimeout = TimeSpan.FromHours(24);

            _hubConnection.On<Ifrit.IfritResult>(HubMethods.IfritRollResultsHub.AddRegularResult, (result) =>
            {
                _results.Push(new ResultItem(result));
                StateHasChanged();
            });

            _hubConnection.On<string, Ifrit.CriticalResult>(HubMethods.IfritRollResultsHub.AddCriticalResult, (name, result) =>
            {
                _results.Push(new ResultItem(name, result));
                StateHasChanged();
            });
```

Let me know if there's anything I can do to try to get this connection to remain stable - 

Thanks!

**Comment by BrennanConroy** [comment]
SignalR connections are designed to be open for long periods of time without any input from the user or server.

Remove the timeout you set as that doesn't do what you want.

Please provide client and server logs so we can see why the client disconnected.
https://docs.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-3.1#server-side-logging
https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-model-configuration?view=aspnetcore-3.1#configure-the-signalr-client-for-blazor-server-apps

**Comment by mandalorianbob** [comment]
Okay I put in the logging but I can't repo it locally, only when I'm online. I got the server log stream going, but where do I get the client logs?

**Comment by BrennanConroy** [comment]
From the browser console.

**Comment by mandalorianbob** [comment]
I found a potential problem in my reconnect logic, and I'm testing that out - but meanwhile
I'm having difficulty turning on the logging - that link was for server side, and from the .net client, turning on the logging doesn't seem to be emitting anything from the console.

I've got this
```
     _hubConnection = new HubConnectionBuilder()
                 .WithUrl(NavigationManager.ToAbsoluteUri($"/{HubMethods.StarWarsRollResultsHub.Name}"))
                 .ConfigureLogging((builder) => builder.SetMinimumLevel(LogLevel.Debug))
                 .WithAutomaticReconnect()
                 .Build();
```

But that doesn't emit anything.

I added
```
  builder.Logging.SetMinimumLevel(LogLevel.Debug);
```
in Program.Main but that seems to only emit Blazor stuff.
[corvustest.azurewebsites.net-1591667684349.log](https://github.com/dotnet/aspnetcore/files/4749511/corvustest.azurewebsites.net-1591667684349.log)


I've attached the server log, but it doesn't seem to have any signalr output. Here's what I put in Program.cs
```
  public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseConfiguration(new ConfigurationBuilder()
                    .AddCommandLine(args)
                    .Build())
                .ConfigureLogging(logging =>
                {
                    logging.AddFilter("Microsoft.AspNetCore.SignalR", LogLevel.Debug);
                    logging.AddFilter("Microsoft.AspNetCore.Http.Connections", LogLevel.Debug);
                })
                .UseStartup<Startup>()
                .Build();
```
[corvusserverlog.txt](https://github.com/dotnet/aspnetcore/files/4749497/corvusserverlog.txt)

**Comment by BrennanConroy** [comment]
Oh right, you're using the SignalR Client with WASM
https://docs.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-3.1#console-logging

Also, the server logs you provided are not the server logs. Try https://docs.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-3.1#azure-app-service

**Comment by mandalorianbob** [comment]
Well, this has been more of a struggle than I expected. I figured out how to fix my reconnect, so I'm not blocked ... but I'm still randomly getting disconnected, and that's probably worth figuring out.

But this logging thing has proven to be more hassle than I expected.

So I installed the Microsoft.Extensions.Logging.Console package and added 
```
 .ConfigureLogging((builder) =>
                 {
                     builder.AddConsole();
                     builder.SetMinimumLevel(LogLevel.Debug);
                 })
```
to my HubConnectionBuilder.

But the page won't load - 

﻿System.InvalidOperationException: A suitable constructor for type 'Microsoft.Extensions.Logging.Console.ConsoleLoggerProvider' could not be located. Ensure the type is concrete and services are registered for all parameters of a public constructor.

I upped the package to 5.0.0-preview.4 just in case, but no joy.  

It feels like I'm missing a package or a service call or something somewhere.

**Comment by BrennanConroy** [comment]
That error is from the browser? Try rebuilding the project and force refreshing the browser page.

**Comment by mandalorianbob** [comment]
I did a clean, deleted the obj/bin folders outta the Client project just to be extra sure, and now I get a different error. 

Since I thought - maybe this is the preview stuff - I removed those nuget packages, downgraded Extensions Logging Console to 3.1.4, cleaned again, and rebuilt, and ran it, and I get the different error again.

![newerror](https://user-images.githubusercontent.com/19713606/84103835-feb43e00-a9c8-11ea-8c39-1e25d7c8052c.png)

**Comment by BrennanConroy** [comment]
@pranavkm Could you help out?

**Comment by pranavkm** [comment]
> I'm using Blazor.HttpClient 3.2.0-preview3.20168.3

The package was removed during preview4. I'll recommend following the upgrade instructions listed in https://devblogs.microsoft.com/aspnet/blazor-webassembly-3-2-0-preview-4-release-now-available/ and continue from there.

The exception message in the most recent stack trace looks like https://github.com/dotnet/runtime/issues/37601, but it's not clear what's spinning up a new thread. Regardless, it's not going to work in wasm 3.2.

**Comment by davidfowl** [comment]
@BrennanConroy @pranavkm  we should make sure errors show up by default . We need to look into this

**Comment by mandalorianbob** [comment]
Good to know about that package. I removed it and replaced it with the recommended System.Net.Http.Json

Now, I stripped out all the stuff I was trying to do to tweak things, and got it back to the basics.

I have Microsoft.Extensions.Logging.Console package 3.1.4

I have a connection that looks like this
```
_hubConnection = new HubConnectionBuilder()
                 .WithUrl(NavigationManager.ToAbsoluteUri($"/{HubMethods.IfritRollResultsHub.Name}"))
                 .ConfigureLogging((builder) =>
                 {
                     //builder.AddConsole();
                     builder.SetMinimumLevel(LogLevel.Debug);
                 })
                 .WithAutomaticReconnect()
                 .Build();
```

And it works, and loads the page. If I uncomment builder.AddConsole, it throws the above error.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. This is a dupe of the referenced issue: https://github.com/dotnet/aspnetcore/issues/22706

**Comment by mandalorianbob** [comment]
No, no it's not. The lack of logging wasn't the issue the random disconnects were the issue.

**Comment by mandalorianbob** [comment]
The logging was to try to figure out the disconnects hahaha

**Comment by mandalorianbob** [comment]
Okay in that other thread though, user got the logging working, I'll get some logs and get back to this thread.

**Comment by mandalorianbob** [comment]
Okay I've got two logs - it looks like what's happening is that the appservice is suddenly running - sandboxproc.exe C:\DWASFiles\Sites\CorvusTest\Temp\applicationhost.config True True
After it does that, the client gets 
blazor.webassembly.js:1 info: Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport[11]
      WebSocket closed by the server. Close status NormalClosure.

And then reconnects. I wasn't handling reconnects properly before, but am now.

I realized the other day I was in debug mode but thought nothing of it - could that be related?

The log is from my App Service Application Logs Log Stream. I tried the Web Server logs but those didn't seem to output anything.


[appservice-applogstream.txt](https://github.com/dotnet/aspnetcore/files/4755828/appservice-applogstream.txt)
[consolelog.txt](https://github.com/dotnet/aspnetcore/files/4755829/consolelog.txt)
